src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -71,6 +94,325 @@ private static Vector128<float> VectorSum(in Vector128<float> vector)
             }
         }
 
+        // Multiply matrix times vector into vector.
+        internal static unsafe void MatMulA(bool add, AlignedArray mat, AlignedArray src, AlignedArray dst, int crow, int ccol)
+        {
+            Contracts.Assert(Compat(mat));
+            Contracts.Assert(Compat(src));
+            Contracts.Assert(Compat(dst));
+
+            fixed (float* pSrcStart = &src.Items[0])
+            fixed (float* pDstStart = &dst.Items[0])
+            fixed (float* pMatStart = &mat.Items[0])
+            {
+                float* psrc = Ptr(src, pSrcStart);
+                float* pdst = Ptr(dst, pDstStart);
+                float* pmat = Ptr(mat, pMatStart);
+
+                float* pSrcEnd = psrc + ccol;
+                float* pDstEnd = pdst + crow;
+                float* pSrcCurrent = psrc;
+                float* pDstCurrent = pdst;
+                float* pMatCurrent = pmat;
+
+                while (pDstCurrent < pDstEnd)
+                {
+                    Vector128<float> res0 = Sse.SetZeroVector128();
+                    Vector128<float> res1 = res0;
+                    Vector128<float> res2 = res0;
+                    Vector128<float> res3 = res0;
+
+                    while (pSrcCurrent < pSrcEnd)
+                    {
+                        float* pMatTemp = pMatCurrent;
+
+                        Vector128<float> x01 = Sse.LoadAlignedVector128(pMatTemp);
+                        Vector128<float> x11 = Sse.LoadAlignedVector128(pMatTemp += ccol);
+                        Vector128<float> x21 = Sse.LoadAlignedVector128(pMatTemp += ccol);
+                        Vector128<float> x31 = Sse.LoadAlignedVector128(pMatTemp += ccol);
+                        Vector128<float> x02 = Sse.LoadAlignedVector128(pSrcCurrent);
+
+                        res0 = Sse.Add(res0, Sse.Multiply(x01, x02));
+                        res1 = Sse.Add(res1, Sse.Multiply(x11, x02));
+                        res2 = Sse.Add(res2, Sse.Multiply(x21, x02));
+                        res3 = Sse.Add(res3, Sse.Multiply(x31, x02));
+
+                        pSrcCurrent += 4;
+                        pMatCurrent += 4;
+                    }
+
+                    // Add up the entries of each, with the 4 results in res0
+                    res0 = Sse3.HorizontalAdd(res0, res1);
+                    res2 = Sse3.HorizontalAdd(res2, res3);
+                    res0 = Sse3.HorizontalAdd(res0, res2);
+
+                    if (add)
+                    {
+                        res0 = Sse.Add(res0, Sse.LoadAlignedVector128(pDstCurrent));
+                    }
+                    Sse.StoreAligned(pDstCurrent, res0);
+
+                    pDstCurrent += 4;
+                    pMatCurrent += 3 * ccol;
```

Comment:
Did you figure out why this is `3 * ccol`?

Reply:
Maybe because here it is adding the ccol 3 times to pMatTemp which is a copy of pMatCurrent and at the end it needs to reset pMatCurrent to the right position?

https://github.com/dotnet/machinelearning/pull/668/files/02bfbe6c0f2f8d12ef7ef37006557d4f0616a054#diff-1212867e04546a8e692d4e263dcc390cR130

Reply:
It is following the same pattern in this method: 

https://github.com/dotnet/machinelearning/pull/668/files/02bfbe6c0f2f8d12ef7ef37006557d4f0616a054#diff-1212867e04546a8e692d4e263dcc390cR292

Reply:
Agree and thanks!  We may get a slightly better view from the native code here:
https://github.com/dotnet/machinelearning/blob/d0664c1063539805c390a0f95231dcbfb87bc974/src/Native/CpuMathNative/Sse.cpp#L108-L146

`pMatCurrent` corresponds to `pm` in the native code in the window.  For correctness, `pm` needs to be incremented by `4 * ccol` in each iteration of the outer for loop.  We see that at the end of such iteration, `pm` is incremented by `3 * ccol` only - this is because the inner for loop increments `pm` by `4` in each of its own iteration.  SInce the inner for loop runs for `(psLim - psrc) / 4 = ccol / 4` iterations, the entire inner for loop increments `pm` by `(ccol / 4) * 4 = ccol`.  Together with the increment by `3 * ccol` done by the outer for loop, `pm` is incremented by `4 * ccol` in each iteration of the outer for loop, which is the desired behavior.